src/types/quota.rs

```diff
@@ -0,0 +1,160 @@
+                        Ok((_rest, Response::Quota(a))) => {
+                            return Ok(a);
+                        }
+                        Ok((rest, data)) => {
+                            lines = rest;
+                            if let Some(resp) = try_handle_unilateral(data, unsolicited) {
+                                return Err(resp.into());
+                            }
+                        }
+                        _ => {
+                            return Err(Error::Parse(ParseError::Invalid(lines.to_vec())));
+                        }
+                    }
+                }
+
+                Err(Error::Parse(ParseError::Invalid(lines.to_vec())))
+            },
+        }
+        .try_build()
+    }
+
+    pub fn root_name(&self) -> &str {
+        &*self.borrow_quota().root_name
+    }
+
+    pub fn resources(&self) -> &[imap_proto::QuotaResource<'_>] {
+        &*self.borrow_quota().resources
+    }
+}
+
+#[self_referencing]
+pub struct QuotaRoot {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) inner: InnerQuotaRoot<'this>,
+}
+
+pub(crate) struct InnerQuotaRoot<'a> {
+    pub(crate) quota_root: imap_proto::QuotaRoot<'a>,
+    pub(crate) quotas: Vec<imap_proto::Quota<'a>>,
+}
+
+impl QuotaRoot {
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        QuotaRootTryBuilder {
+            data: owned,
+            inner_builder: |input| {
+                let mut quota_roots: Vec<imap_proto::QuotaRoot<'_>> = Vec::new();
+                let mut quotas = Vec::new();
+                parse_many_into(
+                    input,
+                    &mut quota_roots,
+                    unsolicited,
+                    |response| match response {
+                        Response::QuotaRoot(q) => Ok(MapOrNot::Map(q)),
+                        Response::Quota(_) => Ok(MapOrNot::Ignore),
+                        resp => Ok(MapOrNot::Not(resp)),
+                    },
+                )?;
+                parse_many_into(input, &mut quotas, unsolicited, |response| match response {
+                    Response::QuotaRoot(_) => Ok(MapOrNot::Ignore),
+                    Response::Quota(q) => Ok(MapOrNot::Map(q)),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })?;
+                if quota_roots.is_empty() {
+                    Err(Error::Parse(ParseError::Invalid(input.to_vec())))
+                } else {
+                    Ok(InnerQuotaRoot {
+                        quota_root: quota_roots.first().unwrap().clone(),
+                        quotas,
+                    })
+                }
+            },
+        }
+        .try_build()
+    }
+
+    pub fn mailbox_name(&self) -> &str {
+        &*self.borrow_inner().quota_root.mailbox_name
+    }
+
+    pub fn quota_root_names(&self) -> Vec<&str> {
```

Comment:
Let's have this return `impl Iterator<Item = &str>` instead. No need for us to dictate the data structure to use.